crates/miden-lib/asm/miden/auth/multisig.masm

```diff
@@ -0,0 +1,284 @@
+use.miden::account
+use.miden::auth
+use.miden::tx
+use.std::crypto::dsa::rpo_falcon512
+use.std::crypto::hashes::rpo
+use.std::word
+
+# CONSTANTS
+# =================================================================================================
+
+# Auth Request Constants
+
+# The event to request an authentication signature.
+const.AUTH_REQUEST=131087
+
+# The event emitted when a signature is not found for a required signer.
+const.UNAUTHORIZED_EVENT=131102
+
+# Storage Layout Constants
+#
+# ┌─────────────────────────────┬──────────┬──────────────┐
+# │    THRESHOLD & APPROVERS    │ PUB KEYS │ EXECUTED TXS │
+# │           (slot)            │   (map)  │    (map)     │
+# ├─────────────────────────────┼──────────┼──────────────┤
+# │              0              │    1     │      2       │
+# └─────────────────────────────┴──────────┴──────────────┘
+
+# The slot in this component's storage layout where both the signature threshold
+# and number of approvers are stored as [threshold, num_approvers, 0, 0].
+# The threshold is guaranteed to be less than or equal to num_approvers.
+const.THRESHOLD_CONFIG_SLOT=0
+
+# The slot in this component's storage layout where the public keys map is stored.
+# Map entries: [key_index, 0, 0, 0] => owner_public_key
+const.PUBLIC_KEYS_MAP_SLOT=1
+
+# The slot in this component's storage layout where executed transactions are stored.
+# Map entries: transaction_message => [is_executed, 0, 0, 0]
+const.EXECUTED_TXS_SLOT=2
+
+# Memory Constants
+
+# Local Memory Addresses
+const.NUM_OF_APPROVERS=0
+const.PUB_KEY_MAP_IDX=4
+const.CURRENT_PK=8
+const.SUCCESSFUL_VERIFICATIONS=12
+
+# Executed Transaction Flag Constant
+const.IS_EXECUTED_FLAG=[1, 0, 0, 0]
+
+# ERRORS
+const.ERR_TX_ALREADY_EXECUTED="failed to approve multisig transaction as it was already executed"
+
+#! Check if transaction has already been executed and add it to executed transactions for replay protection.
+#!
+#! Inputs:  [MSG]
+#! Outputs: []
+#!
+#! Panics if:
+#! - the same transaction has already been executed
+proc.assert_new_tx
+    push.IS_EXECUTED_FLAG
+    # => [[0, 0, 0, is_executed], MSG]
+
+    swapw
+    # => [MSG, IS_EXECUTED_FLAG]
+
+    push.EXECUTED_TXS_SLOT
+    # => [index, MSG, IS_EXECUTED_FLAG]
+
+    # Set the key value pair in the map to mark transaction as executed
+    exec.account::set_map_item
+    # => [OLD_MAP_ROOT, [0, 0, 0, is_executed]]
+
+    dropw drop drop drop
+    # => [is_executed]
+
+    assertz.err=ERR_TX_ALREADY_EXECUTED
+    # => []
+end
+
+#! Verify signatures for all required signers in a loop.
+#!
+#! This procedure iterates through the required number of signers, fetches their public keys
+#! from the provided account storage map slot, verifies their signatures against the transaction message,
+#! and returns the number of successfully verified signatures.
+#!
+#! Inputs:  [pub_key_slot_idx, num_of_approvers, MSG]
+#! Outputs: [num_verified_signatures, MSG]
+proc.verify_signatures.16
```

Comment:
I would probably move this procedure to the `miden::auth::rpo_falcon512` module (and we can keep the name `verify_signatures`).